File: sdk/typescript/packages/muzen-sdk/src/local.ts
import { randomUUID } from "node:crypto";

import { MuzenUnsupportedFeatureError } from "./errors.js";
import {
  RunnerProtocolError,
  RunnerStdioClient,
} from "./protocol.js";
import {
  delay,
  parseTimeoutMs,
  throwIfAborted,
  withTimeout,
} from "./review-flow.js";
import {
  isRunnerArtifactExportResult,
  isRunnerArtifactReadResult,
  isRunnerStatusResult,
  isTerminalStatus,
  mapNotification,
  mapRunnerArtifact,
  mapRunnerResult,
  mapRunnerStatus,
  toRunnerStartParams,
} from "./runner-mapping.js";
import { registerReviewCallbacks } from "./runner-callbacks.js";
import { parseReviewSource } from "./sources.js";
import { UnsupportedWorkspaceProfileCollection } from "./unsupported.js";
import {
  unwrapWebhookHttpResponse,
  unwrapWorkerRun,
} from "./wire-validation.js";
import type {
  HostConfiguration,
  ModelProfile,
  ModelProfileInput,
  Muzen,
  MuzenWorkerRun,
  MuzenWorkerRunOnceOptions,
  MuzenWorkers,
  MuzenWorkerStartOptions,
  MuzenWebhookHandler,
  MuzenWebhookProvider,
  MuzenWebhooks,
  MuzenWebhookResponseOptions,
  MuzenWorkspace,
  ProviderProfile,
  ProviderProfileInput,
  ReviewArtifact,
  ReviewArtifactExport,
  ReviewArtifactExportOptions,
  ReviewArtifactReadOptions,
  ReviewCancelOptions,
  ReviewEvent,
  ReviewOptions,
  ReviewResult,
  ReviewSession,
  ReviewSessionSnapshot,
  ReviewSource,
  ReviewSourceLike,
  ReviewStatus,
} from "./types.js";

export class RunnerBackedMuzen implements Muzen {
  private readonly sessions = new Map<string, RunnerBackedReviewSession>();
  readonly webhooks: MuzenWebhooks;
  readonly workers: MuzenWorkers;

  constructor(private readonly runner: RunnerStdioClient) {
    this.webhooks = new RunnerBackedMuzenWebhooks(runner);
    this.workers = new RunnerBackedMuzenWorkers(runner);
  }

  async review(
    sourceLike: ReviewSourceLike,
    options: ReviewOptions = {},
  ): Promise<ReviewSession> {
    return this.createReviewSession({ source: sourceLike, options });
  }

  workspace(id: string): MuzenWorkspace {
    return new RunnerBackedWorkspace(this, id);
  }

  async createReviewSession(input: {
    source: ReviewSourceLike;
    options?: ReviewOptions;
  }): Promise<ReviewSession> {
    throwIfAborted(input.options?.signal);
    const source = parseReviewSource(input.source);
    const reviewId = `review-${randomUUID()}`;
    const events: ReviewEvent[] = [];
    let hookError: Error | undefined;
    const unsubscribe = this.runner.onNotification((notification) => {
      const event = mapNotification(notification);
      if (event && event.reviewId === reviewId) {
        events.push(event);
        try {
          input.options?.hooks?.onEvent?.(event);
        } catch (error) {
          hookError ??= reviewHookError(error);
        }
      }
    });
    const unsubscribeCallbacks = registerReviewCallbacks(
      this.runner,
      input.options ?? {},
    );
    let startSent = false;
    let startSettled = false;
    const stopCancelOnAbort = onAbort(input.options?.signal, () => {
      if (!startSent || startSettled) {
        return;
      }
      void this.runner.request("run.cancel", {
        runId: reviewId,
        reason: "operation aborted",
      }).catch(() => {});
    });
    try {
      throwIfAborted(input.options?.signal);
      startSent = true;
      const runnerResult = await this.runner.request(
        "run.start",
        toRunnerStartParams(reviewId, source, input.options ?? {}),
      );
      startSettled = true;
      throwIfAborted(input.options?.signal);
      if (hookError) {
        throw hookError;
      }
      const result = mapRunnerResult(reviewId, source, runnerResult);
      const review = new RunnerBackedReviewSession(
        this.runner,
        reviewId,
        "completed",
        source,
        events,
        result,
      );
      this.sessions.set(reviewId, review);
      return review;
    } catch (error) {
      if (
        error instanceof RunnerProtocolError &&
        error.kind === "invalid_input"
      ) {
        throw error;
      }
      throw error;
    } finally {
      startSettled = true;
      stopCancelOnAbort();
      unsubscribeCallbacks();
      unsubscribe();
    }
  }

  async resumeReview(id: string): Promise<ReviewSession> {
    const existing = this.sessions.get(id);
    if (!existing) {
      throw new MuzenUnsupportedFeatureError(
        "local resumeReview is process-local for createMuzen(); use createMuzenClient({ baseUrl }).resumeReview(id) for durable review lookup",
      );
    }
    return existing;
  }

  async close(): Promise<void> {
    await this.runner.close();
  }
}

class RunnerBackedWorkspace implements MuzenWorkspace {
  readonly models = new UnsupportedWorkspaceProfileCollection<
    ModelProfileInput,
    ModelProfile
  >("model");
  readonly providers = new UnsupportedWorkspaceProfileCollection<
    ProviderProfileInput,
    ProviderProfile
  >("provider");

  constructor(
    private readonly muzen: Muzen,
    readonly id: string,
  ) {}

  review(
    source: ReviewSourceLike,
    options?: ReviewOptions,
  ): Promise<ReviewSession> {
    return this.muzen.review(source, options);
  }
}

class RunnerBackedMuzenWebhooks implements MuzenWebhooks {
  readonly github: MuzenWebhookHandler;
  readonly gitlab: MuzenWebhookHandler;

  constructor(runner: RunnerStdioClient) {
    this.github = new RunnerBackedMuzenWebhookHandler(runner, "github");
    this.gitlab = new RunnerBackedMuzenWebhookHandler(runner, "gitlab");
  }
}

class RunnerBackedMuzenWebhookHandler implements MuzenWebhookHandler {
  constructor(
    private readonly runner: RunnerStdioClient,
    private readonly provider: MuzenWebhookProvider,
  ) {}

  async response(
    request: Request,
    options: MuzenWebhookResponseOptions = {},
  ): Promise<Response> {
    if (request.method !== "POST") {
      throw new MuzenUnsupportedFeatureError(
        `Muzen ${this.provider} webhook handlers expect POST requests`,
      );
    }
    throwIfAborted(options.signal ?? request.signal);
    const result = unwrapWebhookHttpResponse(
      await this.runner.request(`webhook.${this.provider}.handle`, {
        workspaceId: options.workspaceId,
        headers: headersToRecord(request.headers),
        body: await request.text(),
        secret: webhookSecret(this.provider, options.secret),
        options: {
          reviewOptions: options.review ?? {},
        },
      }),
    );
    return new Response(result.body, {
      status: result.statusCode,
      headers: result.headers,
    });
  }
}

class RunnerBackedMuzenWorkers implements MuzenWorkers {
  constructor(private readonly runner: RunnerStdioClient) {}

  async runOnce(
    options: MuzenWorkerRunOnceOptions = {},
  ): Promise<MuzenWorkerRun> {
    throwIfAborted(options.signal);
    const run = unwrapWorkerRun(
      await this.runner.request("worker.runOnce", {
        workerId: options.workerId,
        maxSessions: options.maxSessions ?? 1,
        hostConfig: options.hostConfig ?? {},
      }),
    );
    throwIfAborted(options.signal);
    return run;
  }

  async start(options: MuzenWorkerStartOptions = {}): Promise<void> {
    validateWorkerQueues(options.queues);
    const maxSessions = options.maxSessions ?? options.concurrency ?? 1;
    const pollIntervalMs = options.pollIntervalMs ?? 1_000;
    while (true) {
      const run = await this.runOnce({
        workerId: options.workerId,
        maxSessions,
        hostConfig: workerHostConfig(options),
        signal: options.signal,
      });
      if (options.stopWhenIdle && run.claimed === 0) {
        return;
      }
      await delay(pollIntervalMs, options.signal);
    }
  }
}

class RunnerBackedReviewSession implements ReviewSession {
  private readonly listeners = new Set<(event: ReviewEvent) => void>();

  constructor(
    private readonly runner: RunnerStdioClient,
    readonly id: string,
    private currentStatus: ReviewStatus,
    readonly source: ReviewSource,
    private readonly recordedEvents: ReviewEvent[],
    private currentResult?: ReviewResult,
  ) {}

  get status(): ReviewStatus {
    return this.currentStatus;
  }

  subscribe(
    listener: (event: ReviewEvent) => void,
    options: { replay?: boolean } = {},
  ): () => void {
    if (options.replay ?? true) {
      for (const event of this.recordedEvents) {
        listener(event);
      }
    }
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  async *events(options: {
    after?: string | null;
    signal?: AbortSignal;
  } = {}): AsyncIterable<ReviewEvent> {
    throwIfAborted(options.signal);
    const start = afterCursorIndex(this.recordedEvents, options.after);
    for (const event of this.recordedEvents.slice(start)) {
      throwIfAborted(options.signal);
      yield event;
    }
  }

  eventsResponse(options: {
    after?: string | null;
    signal?: AbortSignal;
  } = {}): Response {
    const events = this.events(options);
    const encoder = new TextEncoder();
    const stream = new ReadableStream<Uint8Array>({
      async start(controller) {
        try {
          for await (const event of events) {
            controller.enqueue(
              encoder.encode(
                `id: ${event.cursor}\nevent: ${event.type}\ndata: ${JSON.stringify(event)}\n\n`,
              ),
            );
          }
          controller.close();
        } catch (error) {
          controller.error(error);
        }
      },
    });
    return new Response(stream, {
      headers: {
        "Content-Type": "text/event-stream",
      },
    });
  }

  async wait(options: {
    timeout?: string | number;
    signal?: AbortSignal;
  } = {}): Promise<ReviewResult> {
    throwIfAborted(options.signal);
    if (this.currentResult) {
      return this.currentResult;
    }
    const timeoutMs = parseTimeoutMs(options.timeout);
    const resultPromise = this.result().then((result) => {
      if (!result) {
        throw new Error(`review ${this.id} has no final result yet`);
      }
      return result;
    });
    if (timeoutMs === undefined) {
      return resultPromise;
    }
    return withTimeout(resultPromise, timeoutMs);
  }

  async result(): Promise<ReviewResult | undefined> {
    if (this.currentResult) {
      return this.currentResult;
    }
    const runnerResult = await this.runner.request("run.result", {
      runId: this.id,
    });
    this.currentResult = mapRunnerResult(this.id, this.source, runnerResult);
    this.currentStatus = this.currentResult.status;
    return this.currentResult;
  }

  async readArtifact(
    artifactId: string,
    options: ReviewArtifactReadOptions = {},
  ): Promise<ReviewArtifact> {
    const result = await this.runner.request("artifact.read", {
      runId: this.id,
      artifactId,
      view: options.view ?? "redacted",
    });
    if (!isRunnerArtifactReadResult(result)) {
      throw new Error("muzen-runner returned an invalid artifact read result");
    }
    return mapRunnerArtifact(result.artifact);
  }

  async exportArtifacts(
    options: ReviewArtifactExportOptions = {},
  ): Promise<ReviewArtifactExport> {
    const result = await this.runner.request("artifact.export", {
      runId: this.id,
      artifactIds: options.artifactIds ?? [],
      view: options.view ?? "redacted",
      maxArtifacts: options.maxArtifacts,
      maxBytes: options.maxBytes,
    });
    if (!isRunnerArtifactExportResult(result)) {
      throw new Error("muzen-runner returned an invalid artifact export result");
    }
    return {
      view: result.view,
      artifactCount: result.artifactCount,
      totalBytes: result.totalBytes,
      artifacts: result.artifacts.map(mapRunnerArtifact),
    };
  }

  async cancel(reason?: string | ReviewCancelOptions): Promise<void> {
    const cancelReason =
      typeof reason === "string" ? reason : reason?.reason ?? "cancelled";
    await this.runner.request("run.cancel", {
      runId: this.id,
      reason: cancelReason,
    });
    if (!isTerminalStatus(this.currentStatus)) {
      this.currentStatus = "cancelled";
      this.record({
        cursor: String(this.recordedEvents.length + 1),
        type: "session.cancelled",
        reviewId: this.id,
        timestampUtc: new Date().toISOString(),
        payload: { reason: cancelReason },
      });
    }
  }

  async refresh(): Promise<ReviewSessionSnapshot> {
    const status = await this.runner.request("run.status", {
      runId: this.id,
    });
    if (isRunnerStatusResult(status)) {
      this.currentStatus = mapRunnerStatus(status.status);
    }
    return {
      id: this.id,
      status: this.currentStatus,
      source: this.source,
      result: this.currentResult,
    };
  }

  private record(event: ReviewEvent): void {
    this.recordedEvents.push(event);
    for (const listener of this.listeners) {
      listener(event);
    }
  }
}

function afterCursorIndex(
  events: ReviewEvent[],
  after: string | null | undefined,
): number {
  if (!after) {
    return 0;
  }
  const index = events.findIndex((event) => event.cursor === after);
  return index === -1 ? 0 : index + 1;
}

function reviewHookError(error: unknown): Error {
  if (error instanceof Error) {
    return error;
  }
  return new Error(`review event hook failed: ${String(error)}`);
}

function onAbort(signal: AbortSignal | undefined, handler: () => void): () => void {
  if (!signal) {
    return () => {};
  }
  if (signal.aborted) {
    handler();
    return () => {};
  }
  signal.addEventListener("abort", handler, { once: true });
  return () => signal.removeEventListener("abort", handler);
}

function headersToRecord(headers: Headers): Record<string, string> {
  const result: Record<string, string> = {};
  headers.forEach((value, key) => {
    result[key] = value;
  });
  return result;
}

function webhookSecret(
  provider: MuzenWebhookProvider,
  explicit: string | undefined,
): string | undefined {
  if (explicit !== undefined) {
    return explicit;
  }
  if (provider === "github") {
    return process.env.GITHUB_WEBHOOK_SECRET;
  }
  return process.env.GITLAB_WEBHOOK_TOKEN ?? process.env.GITLAB_WEBHOOK_SECRET;
}

function validateWorkerQueues(queues: string[] | undefined): void {
  if (!queues) {
    return;
  }
  for (const queue of queues) {
    if (queue !== "reviews") {
      throw new MuzenUnsupportedFeatureError(
        `unsupported Muzen worker queue: ${queue}`,
      );
    }
  }
}

function workerHostConfig(options: MuzenWorkerStartOptions): HostConfiguration {
  const hostConfig: HostConfiguration = {
    ...(options.hostConfig ?? {}),
  };
  if (options.concurrency !== undefined) {
    hostConfig.scheduling = {
      ...(hostConfig.scheduling ?? {}),
      concurrency: {
        ...(hostConfig.scheduling?.concurrency ?? {}),
        maxRunningGlobal: options.concurrency,
      },
    };
  }
  return hostConfig;
}

File: sdk/python/muzen/local.py
from __future__ import annotations

import asyncio
import os
import uuid
from datetime import datetime, timezone
from typing import Any, AsyncIterator, Callable, Dict, List, Optional, Union

from .errors import MuzenUnsupportedFeatureError
from .runner import RunnerStdioClient
from .runner_mapping import (
    _map_notification,
    _map_runner_artifact,
    _map_runner_result,
    _map_runner_status,
    _map_swarm_result,
    _to_runner_start_params,
    _to_swarm_start_params,
)
from .sources import parse_review_source
from .types import (
    ContextEngineConfig,
    ContextLearningScope,
    ContextLearningSource,
    ContextPackPurpose,
    ContextQueryKind,
    ContextQueryLimits,
    ReviewArtifact,
    ReviewArtifactExport,
    ReviewArtifactExportOptions,
    ReviewArtifactReadOptions,
    ReviewCancelOptions,
    ReviewEvent,
    ReviewOptions,
    ReviewResult,
    ReviewSessionSnapshot,
    ReviewSource,
    ReviewSourceLike,
    ReviewStatus,
    SwarmOptions,
    SwarmResult,
)


class Client:
    def __init__(self, runner: RunnerStdioClient) -> None:
        self._runner = runner
        self._sessions: Dict[str, ReviewSession] = {}

    @classmethod
    async def create(
        cls,
        *,
        runner_path: Optional[str] = None,
        runner_args: Optional[List[str]] = None,
        client_name: str = "muzen-py",
        client_version: Optional[str] = None,
    ) -> "Client":
        runner = await RunnerStdioClient.start(
            runner_path or os.environ.get("MUZEN_RUNNER_PATH", "muzen-runner"),
            runner_args or ["stdio"],
        )
        await runner.handshake(client_name=client_name, client_version=client_version)
        return cls(runner)

    async def review(
        self,
        source_like: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "ReviewSession":
        source = parse_review_source(source_like)
        review_id = f"review-{uuid.uuid4()}"
        events: List[ReviewEvent] = []

        def on_notification(notification: Dict[str, Any]) -> None:
            event = _map_notification(notification)
            if event and event.review_id == review_id:
                events.append(event)

        unsubscribe = self._runner.on_notification(on_notification)
        try:
            runner_result = await self._runner.request(
                "run.start",
                _to_runner_start_params(review_id, source, options or ReviewOptions()),
            )
        finally:
            unsubscribe()
        result = _map_runner_result(review_id, source, runner_result)
        session = ReviewSession(
            runner=self._runner,
            id=review_id,
            status=result.status,
            source=source,
            events=events,
            result=result,
        )
        self._sessions[review_id] = session
        return session

    async def create_review_session(
        self,
        *,
        source: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "ReviewSession":
        return await self.review(source, options)

    async def run_swarm(self, options: SwarmOptions) -> SwarmResult:
        run_id = f"swarm-{uuid.uuid4()}"
        runner_result = await self._runner.request(
            "run.start",
            _to_swarm_start_params(run_id, options),
        )
        return _map_swarm_result(run_id, runner_result)

    async def resume_review(self, review_id: str) -> "ReviewSession":
        try:
            return self._sessions[review_id]
        except KeyError as error:
            raise MuzenUnsupportedFeatureError(
                "local resume_review is process-local for Client.create(); use create_muzen_client(base_url=...).resume_review(review_id) for durable review lookup"
            ) from error

    async def close(self) -> None:
        await self._runner.close()

    def workspace(self, workspace_id: str) -> "RunnerBackedWorkspace":
        return RunnerBackedWorkspace(self, workspace_id)


class RunnerBackedWorkspace:
    def __init__(self, client: Client, workspace_id: str) -> None:
        self._client = client
        self.id = workspace_id
        self.models = UnsupportedWorkspaceProfileCollection("model")
        self.providers = UnsupportedWorkspaceProfileCollection("provider")
        self.context = RunnerBackedContextWorkspace(client._runner)

    async def review(
        self,
        source_like: ReviewSourceLike,
        options: Optional[ReviewOptions] = None,
    ) -> "ReviewSession":
        return await self._client.review(source_like, options)


class RunnerBackedContextWorkspace:
    def __init__(self, runner: RunnerStdioClient) -> None:
        self._runner = runner

    async def index(
        self,
        *,
        source: ReviewSourceLike,
        changed_files: Optional[List[str]] = None,
        config: Optional[ContextEngineConfig] = None,
    ) -> Dict[str, Any]:
        return await self._runner.request(
            "context.index",
            _context_index_params(source, changed_files, config),
        )

    async def build_pack(
        self,
        *,
        source: ReviewSourceLike,
        changed_files: Optional[List[str]] = None,
        purpose: Optional[ContextPackPurpose] = None,
        max_tokens: Optional[int] = None,
        config: Optional[ContextEngineConfig] = None,
    ) -> Dict[str, Any]:
        manifest = await self.index(
            source=source,
            changed_files=changed_files,
            config=config,
        )
        return await self._runner.request(
            "context.pack",
            {
                "snapshotId": manifest["snapshotId"],
                "purpose": purpose or "general_review",
                "maxTokens": max_tokens
                or (config.max_pack_tokens if config is not None else 12_000),
                "seedEvidence": [],
            },
        )

    async def query(
        self,
        *,
        source: ReviewSourceLike,
        kind: ContextQueryKind,
        arguments: Optional[Dict[str, Any]] = None,
        changed_files: Optional[List[str]] = None,
        purpose: Optional[ContextPackPurpose] = None,
        current_evidence: Optional[List[str]] = None,
        limits: Optional[ContextQueryLimits] = None,
        config: Optional[ContextEngineConfig] = None,
    ) -> Dict[str, Any]:
        manifest = await self.index(
            source=source,
            changed_files=changed_files,
            config=config,
        )
        runner_limits = (
            {
                "maxResults": limits.max_results,
                "maxTokens": limits.max_tokens,
            }
            if limits is not None
            else {
                "maxResults": config.max_query_results if config is not None else 120,
                "maxTokens": config.max_pack_tokens if config is not None else 12_000,
            }
        )
        return await self._runner.request(
            "context.query",
            {
                "snapshotId": manifest["snapshotId"],
                "purpose": purpose,
                "kind": kind,
                "arguments": arguments or {},
                "currentEvidence": current_evidence or [],
                "limits": runner_limits,
            },
        )

    async def record_feedback(
        self,
        *,
        source: ReviewSourceLike,
        feedback: str,
        evidence_ids: Optional[List[str]] = None,
        changed_files: Optional[List[str]] = None,
        learning_source: Optional[ContextLearningSource] = None,
        scope: Optional[ContextLearningScope] = None,
        config: Optional[ContextEngineConfig] = None,
    ) -> Dict[str, Any]:
        manifest = await self.index(
            source=source,
            changed_files=changed_files,
            config=config,
        )
        return await self._runner.request(
            "context.feedback",
            {
                "snapshotId": manifest["snapshotId"],
                "evidenceIds": evidence_ids or [],
                "feedback": feedback,
                "source": learning_source,
                "scope": scope,
            },
        )

    async def approve_learning(
        self,
        *,
        snapshot_id: str,
        learning_id: str,
        approve: bool = False,
        expires_at_utc: Optional[str] = None,
    ) -> Dict[str, Any]:
        return await self._runner.request(
            "context.learning.approve",
            {
                "snapshotId": snapshot_id,
                "learningId": learning_id,
                "approve": approve,
                "expiresAtUtc": expires_at_utc,
            },
        )


def _context_index_params(
    source_like: ReviewSourceLike,
    changed_files: Optional[List[str]],
    config: Optional[ContextEngineConfig],
) -> Dict[str, Any]:
    source = parse_review_source(source_like)
    if source.type == "local":
        repo = source.repo
        default_changed_files = source.changed_files or []
    elif source.type == "raw_snapshot":
        repo = source.root
        default_changed_files = source.changed_files or []
    else:
        raise MuzenUnsupportedFeatureError(
            "local context methods require a local or raw_snapshot source"
        )
    payload: Dict[str, Any] = {
        "repo": repo,
        "changedFiles": changed_files if changed_files is not None else default_changed_files,
    }
    if config is not None:
        payload["config"] = {
            "mode": config.mode,
            "maxIndexedFiles": config.max_indexed_files,
            "maxIndexedBytes": config.max_indexed_bytes,
            "maxEvidenceItems": config.max_evidence_items,
            "maxPackTokens": config.max_pack_tokens,
            "maxQueryResults": config.max_query_results,
            "includeRepositoryGuidance": config.include_repository_guidance,
            "includeHostContext": config.include_host_context,
            "strictEvidenceRequired": config.strict_evidence_required,
        }
    return payload


class UnsupportedWorkspaceProfileCollection:
    def __init__(self, kind: str) -> None:
        self._kind = kind

    async def set(self, name: str, input: Any) -> Any:
        raise self._error()

    async def get(self, name: str) -> Any:
        raise self._error()

    async def list(self) -> List[Any]:
        raise self._error()

    def _error(self) -> MuzenUnsupportedFeatureError:
        return MuzenUnsupportedFeatureError(
            f"workspace {self._kind} profiles require remote workspace storage; Client.create() only supports local runner review execution in this preview"
        )


class ReviewSession:
    def __init__(
        self,
        *,
        runner: RunnerStdioClient,
        id: str,
        status: ReviewStatus,
        source: ReviewSource,
        events: List[ReviewEvent],
        result: Optional[ReviewResult],
    ) -> None:
        self._runner = runner
        self.id = id
        self.status = status
        self.source = source
        self._events = events
        self._result = result
        self._listeners: List[Callable[[ReviewEvent], None]] = []

    def subscribe(
        self,
        listener: Callable[[ReviewEvent], None],
        *,
        replay: bool = True,
    ) -> Callable[[], None]:
        if replay:
            for event in self._events:
                listener(event)
        self._listeners.append(listener)

        def unsubscribe() -> None:
            if listener in self._listeners:
                self._listeners.remove(listener)

        return unsubscribe

    async def events(self, *, after: Optional[str] = None) -> AsyncIterator[ReviewEvent]:
        start = _after_cursor_index(self._events, after)
        for event in self._events[start:]:
            yield event

    async def wait(
        self,
        *,
        timeout: Optional[Union[int, float, str]] = None,
    ) -> ReviewResult:
        if self._result is not None:
            return self._result
        timeout_seconds = _parse_timeout_seconds(timeout)
        result_task = asyncio.create_task(self.result())
        result = await asyncio.wait_for(result_task, timeout=timeout_seconds)
        if result is None:
            raise RuntimeError(f"review {self.id} has no final result yet")
        return result

    async def result(self) -> Optional[ReviewResult]:
        if self._result is not None:
            return self._result
        runner_result = await self._runner.request("run.result", {"runId": self.id})
        self._result = _map_runner_result(self.id, self.source, runner_result)
        self.status = self._result.status
        return self._result

    async def read_artifact(
        self,
        artifact_id: str,
        options: Optional[ReviewArtifactReadOptions] = None,
    ) -> ReviewArtifact:
        options = options or ReviewArtifactReadOptions()
        result = await self._runner.request(
            "artifact.read",
            {
                "runId": self.id,
                "artifactId": artifact_id,
                "view": options.view,
            },
        )
        if not isinstance(result, dict) or not isinstance(result.get("artifact"), dict):
            raise RuntimeError("muzen-runner returned an invalid artifact read result")
        return _map_runner_artifact(result["artifact"])

    async def export_artifacts(
        self,
        options: Optional[ReviewArtifactExportOptions] = None,
    ) -> ReviewArtifactExport:
        options = options or ReviewArtifactExportOptions()
        result = await self._runner.request(
            "artifact.export",
            {
                "runId": self.id,
                "artifactIds": options.artifact_ids,
                "view": options.view,
                "maxArtifacts": options.max_artifacts,
                "maxBytes": options.max_bytes,
            },
        )
        if not isinstance(result, dict):
            raise RuntimeError("muzen-runner returned an invalid artifact export result")
        return ReviewArtifactExport(
            view=result.get("view", options.view),
            artifact_count=result.get("artifactCount", 0),
            total_bytes=result.get("totalBytes", 0),
            artifacts=[
                _map_runner_artifact(artifact)
                for artifact in result.get("artifacts", [])
                if isinstance(artifact, dict)
            ],
        )

    async def cancel(
        self,
        reason: Optional[Union[str, ReviewCancelOptions]] = None,
    ) -> None:
        cancel_reason = reason.reason if isinstance(reason, ReviewCancelOptions) else reason
        await self._runner.request(
            "run.cancel",
            {"runId": self.id, "reason": cancel_reason or "cancelled"},
        )
        if self.status not in ("completed", "failed", "cancelled"):
            self.status = "cancelled"
            self._record(
                ReviewEvent(
                    cursor=str(len(self._events) + 1),
                    type="session.cancelled",
                    review_id=self.id,
                    timestamp_utc=_timestamp_utc(),
                    payload={"reason": cancel_reason},
                )
            )

    async def refresh(self) -> ReviewSessionSnapshot:
        status = await self._runner.request("run.status", {"runId": self.id})
        if isinstance(status, dict) and isinstance(status.get("status"), str):
            self.status = _map_runner_status(status["status"])
        return ReviewSessionSnapshot(
            id=self.id,
            status=self.status,
            source=self.source,
            result=self._result,
        )

    def _record(self, event: ReviewEvent) -> None:
        self._events.append(event)
        for listener in list(self._listeners):
            listener(event)


def _after_cursor_index(events: List[ReviewEvent], after: Optional[str]) -> int:
    if not after:
        return 0
    for index, event in enumerate(events):
        if event.cursor == after:
            return index + 1
    return 0


def _parse_timeout_seconds(timeout: Optional[Union[int, float, str]]) -> Optional[float]:
    if timeout is None:
        return None
    if isinstance(timeout, (int, float)):
        return float(timeout) / 1000
    text = timeout.strip()
    if text.endswith("ms"):
        return float(text[:-2]) / 1000
    if text.endswith("s"):
        return float(text[:-1])
    if text.endswith("m"):
        return float(text[:-1]) * 60
    return float(text) / 1000


def _timestamp_utc() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
